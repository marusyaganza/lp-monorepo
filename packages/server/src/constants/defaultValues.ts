import { Language, Role, SortWordsBy, Tense } from '../generated/graphql';

export const WORDS_PER_PAGE = 5;
export const DEFAULT_LANGUAGE = Language.English;
export const DEFAULT_WORDS_SORT_BY = SortWordsBy.UpdatedAt;
export const DEFAULT_ROLE = Role.Member;
export const DEFAULT_TENSE = Tense.Pind;
export const DEFAULT_TOKEN_TTL = 7;
// to clean db every hour + 10 min just in case
export const DEMO_DB_TTL = 4200;
export const DEFAULT_MONGO_CONNECTION = 'mongodb://localhost:27017';
export const DEFAULT_OPTIONS__SAMPLE_SIZE = 50;
