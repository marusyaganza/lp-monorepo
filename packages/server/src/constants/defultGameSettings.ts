import { Game, Language, SettingsByGame } from '../generated/graphql';

export const DEFAULT_GAMES_SETTINGS: SettingsByGame = {
  [Game.Audio]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Image]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Speaking]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.TypeWord]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.SelectDef]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 6,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.SelectWord]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 6,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Conjugation]: {
    wordsPerGame: 2,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.Spanish]
  },
  [Game.Gender]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.Spanish]
  }
};
