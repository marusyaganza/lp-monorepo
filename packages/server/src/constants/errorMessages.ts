export const ERROR_MESSAGES = {
  NOT_AUTHENTICATED: 'Login to perform this operation',
  NOT_AUTHORIZED: 'You do not have permission to perform this operation',
  INCORRECT_CREDS: 'email or password is incorrect',
  SIGN_UP_FAILED: 'sign up operation failed',
  SING_UP_UNAVAILABLE:
    'Sign up operation is not available now. Sorry for inconvenience.',
  GAME_NOT_FOUND: 'Game not found',
  GAME_GENERATION_FAILED: 'Game data generation failed'
};
