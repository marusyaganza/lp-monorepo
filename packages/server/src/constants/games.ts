import { Game, GameConfig, Language } from '../generated/graphql';
export const GAMES: GameConfig[] = [
  {
    name: 'WordEcho',
    id: 'audio',
    desc: 'Listen, Type, Master. Sharpen language skills through auditory engagement.',
    type: Game.Audio,
    orderNum: 0,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'LexiSense',
    desc: 'Define & Conquer.Match words with their meanings intuitively.',
    id: 'select_definition',
    type: Game.SelectDef,
    orderNum: 1,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'Definify: Match & Learn',
    desc: 'Associate definitions with the right words. A captivating word puzzle adventure!',
    id: 'select_word',
    type: Game.SelectWord,
    orderNum: 2,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'WordRecall: Define & Type',
    desc: 'Type words matching given definitions. Boost vocabulary in engaging rounds!',
    id: 'type_word',
    type: Game.TypeWord,
    orderNum: 3,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'PictureWord Blitz',
    desc: 'Identify objects from pictures and type their names to enhance vocabulary skills!',
    id: 'image',
    type: Game.Image,
    orderNum: 4,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'VoiceWord Challenge',
    desc: 'Speak words from clues and images. Sharpen your vocabulary and speaking skills fast!',
    id: 'speaking',
    type: Game.Speaking,
    orderNum: 5,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'Conjugation Quest',
    desc: 'Master verb conjugations across tenses and languages in fast-paced, interactive challenges!',
    id: 'conjugation',
    type: Game.Conjugation,
    orderNum: 6,
    languages: [Language.Spanish]
  },
  {
    name: 'GenderWord Challenge',
    desc: 'Choose nouns and match their gender. Test linguistic skills in thrilling rounds!',
    id: 'gender',
    type: Game.Gender,
    orderNum: 7,
    languages: [Language.Spanish]
  }
];
