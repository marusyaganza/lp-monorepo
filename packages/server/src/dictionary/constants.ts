import { Tense } from 'generated/graphql';

export type TagType = {
  tag: string;
  replacement: string;
};

export type ComplexTagType = {
  opening: string;
  closing: string;
};

export const TAGS: TagType[] = [
  { tag: '{b}', replacement: '<b>' },
  { tag: '{/b}', replacement: '</b>' },
  { tag: '{ldquo}', replacement: '\u201C' },
  { tag: '{rdquo}', replacement: '\u201D' },
  { tag: '{it}', replacement: '<i>' },
  { tag: '{/it}', replacement: '</i>' },
  { tag: '{wi}', replacement: '<i>' },
  { tag: '{/wi}', replacement: '</i>' },
  { tag: '{gloss}=', replacement: '<i>' },
  { tag: '{/gloss}', replacement: '</i>' },
  { tag: '{sc}', replacement: '<i>' },
  { tag: '{/sc}', replacement: '</i>' },
  { tag: '{parahw}', replacement: '</i>' },
  { tag: '{/parahw}', replacement: '</i>' },
  { tag: '{phrase}', replacement: '</i>' },
  { tag: '{/phrase}', replacement: '</i>' },
  { tag: '{qword}', replacement: '</i>' },
  { tag: '{/qword}', replacement: '</i>' },
  { tag: '{itsc}', replacement: '<i>' },
  { tag: '{/itsc}', replacement: '</i>' },
  { tag: '{inf}', replacement: '' },
  { tag: '{/inf}', replacement: '' },
  { tag: '{sup}', replacement: '' },
  { tag: '{/sup}', replacement: '' },
  { tag: '{bc}', replacement: '' },
  { tag: '{dx}', replacement: '' },
  { tag: '{/dx}', replacement: '' },
  { tag: '{dx_def}', replacement: '' },
  { tag: '{/dx_def}', replacement: '' },
  { tag: '{dx_ety}', replacement: '' },
  { tag: '{/dx_ety}', replacement: '' },
  { tag: '{ma}', replacement: '' },
  { tag: '{/ma}', replacement: '' },
  { tag: '{rom}', replacement: '' },
  { tag: '{/rom}', replacement: '' }
];

export const COMPLEX_TAGS: ComplexTagType[] = [
  { opening: '{sx', closing: '}' },
  { opening: '{d_link', closing: '}' },
  { opening: '{a_link', closing: '}' },
  { opening: '{i_link', closing: '}' },
  { opening: '{et_link', closing: '}' },
  { opening: '{mat', closing: '}' },
  { opening: '{dxt', closing: '}' }
];

export const COMPLEX_TAGS_TO_REMOVE: ComplexTagType[] = [
  { opening: '{ds|', closing: '}' }
];

export const HW_TAGS: TagType[] = [
  { tag: '{bit}', replacement: '' },
  { tag: '{/bit}', replacement: '' }
];

// Uncomment more tenses as needed
export const TENSES: Record<Tense, string> = {
  pind: 'Presente indicativo',
  pprf: 'Pretérito perfecto simple',
  impf: 'Imperativo',
  pret: 'Pretérito imperfecto'
  //   futr: 'Futuro indicativo',
  //   cond: 'Condicional',
  //   psub: 'Presente subjentivo',
  //   pisb1: 'Pretérito imperfecto',
  //   pisb2: 'Pretérito imperfecto 2',
  //   fsub: 'Futuro',
  //   ppci: 'Pretérito perfecto compuesto',
  //   ppsi: 'Pretérito pluscuamperfecto',
  //   pant: 'Pretérito anterior',
  //   fpin: 'Futuro perfecto',
  //   cpef: 'Condicional perfecto',
  //   ppfs: 'Pretérito perfecto',
  //   ppss1: 'Pretérito pluscuamperfecto',
  //   ppss2: 'Pretérito pluscuamperfecto 2',
  //   fpsb: 'Futuro perfecto',
};
