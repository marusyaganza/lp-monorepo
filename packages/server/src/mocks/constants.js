export const mockEnv = {
  SALT_NUMBER: 3,
  JWT_SECTET: 'JWT_secret'
};
