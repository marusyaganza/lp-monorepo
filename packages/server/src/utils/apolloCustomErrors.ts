import { GraphQLError } from 'graphql';
import { ApolloServerErrorCode } from '@apollo/server/errors';
import { ERROR_MESSAGES } from '../constants/errorMessages';

export enum ERROR_CODES {
  Unauthorized = 'UNAUTHORIZED',
  Forbidden = 'FORBISSEN'
}

export class AuthenticationError extends GraphQLError {
  constructor(errorMessage = ERROR_MESSAGES.NOT_AUTHENTICATED) {
    super(errorMessage, { extensions: { code: ERROR_CODES.Unauthorized } });
  }
}

export class AuthorizationError extends GraphQLError {
  constructor(errorMessage = ERROR_MESSAGES.NOT_AUTHORIZED) {
    super(errorMessage, { extensions: { code: ERROR_CODES.Forbidden } });
  }
}

export class UserInputError extends GraphQLError {
  constructor(errorMessage: string) {
    super(errorMessage, {
      extensions: { code: ApolloServerErrorCode.BAD_USER_INPUT }
    });
  }
}

export class OperationResolutionError extends GraphQLError {
  constructor(errorMessage: string) {
    super(errorMessage, {
      extensions: {
        code: ApolloServerErrorCode.OPERATION_RESOLUTION_FAILURE
      }
    });
  }
}
