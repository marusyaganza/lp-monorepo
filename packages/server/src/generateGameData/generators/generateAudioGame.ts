import { Game, GameData, GameQuestion, Word } from '../../generated/graphql';
import { getExamples, getGameTask } from '../helpers';
import { isWordWithAudio } from '../../types/typeGuards';
import { OperationResolutionError } from '../../utils/apolloCustomErrors';
import { ERROR_MESSAGES } from '../../constants/errorMessages';

export async function generateAudioGame(words: Word[]): Promise<GameData> {
  const gameType = Game.Audio;

  const questions = words.map(word => {
    if (!isWordWithAudio(word)) {
      throw new OperationResolutionError(ERROR_MESSAGES.GAME_GENERATION_FAILED);
    }
    const {
      name,
      audioUrl,
      id,
      defs,
      imgUrl,
      shortDef,
      alternativeSpelling = []
    } = word;

    const examples = getExamples(defs);

    const answer = [name, ...alternativeSpelling];

    const result: GameQuestion = {
      answer: answer,
      question: [audioUrl],
      wordId: id,
      additionalInfo: {
        imgUrl,
        examples,
        shortDef: `<b>${name} means</b> ${shortDef[0]}`
      }
    };

    return result;
  });

  const task = getGameTask(gameType);

  return { questions, task, type: gameType };
}
