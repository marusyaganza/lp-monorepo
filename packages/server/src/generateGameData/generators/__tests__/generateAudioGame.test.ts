import { testData } from '../../../tests/mocks/wordsForGames';
import { generateAudioGame } from '../generateAudioGame';
import { Game, Language } from '../../../generated/graphql';
import { OperationResolutionError } from '../../../utils/apolloCustomErrors';
import { ERROR_MESSAGES } from '../../../constants/errorMessages';
import { DEFAULT_GAMES_SETTINGS } from '../../../constants/defaultGameSettings';

const languages = Object.values(Language);
const gameType = Game.Audio;

const gameConfig = DEFAULT_GAMES_SETTINGS[gameType];
describe('generateAudioGame', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  languages.forEach(language => {
    it(`should return the correct number of questions when data has enough ${language} words`, async () => {
      const gameData = await generateAudioGame(testData[language]);
      expect(gameData?.questions?.length).toBe(gameConfig?.wordsPerGame);
      expect(gameData.type).toBe(gameType);
      expect(gameData).toMatchSnapshot();
    });

    it(`should return the correct number of questions when with minimal ${language} words`, async () => {
      const gameData = await generateAudioGame([testData[language][0]]);
      expect(gameData?.questions?.length).toBe(gameConfig?.minWords);
      expect(gameData.type).toBe(gameType);
      expect(gameData).toMatchSnapshot();
    });
  });
  it('should throw an error with incorrect data', async () => {
    let error;
    try {
      // @ts-expect-error: using incorrect arguments here
      await generateAudioGame([{}]);
    } catch (err) {
      error = err;
    }
    expect(error).toEqual(
      new OperationResolutionError(ERROR_MESSAGES.GAME_GENERATION_FAILED)
    );
  });
});
