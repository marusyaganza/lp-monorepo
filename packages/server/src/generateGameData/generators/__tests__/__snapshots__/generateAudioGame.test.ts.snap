// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateAudioGame should return the correct number of questions when data has enough ENGLISH words 1`] = `
{
  "questions": [
    {
      "additionalInfo": {
        "examples": [],
        "imgUrl": "https://example.com/images/apple.jpg",
        "shortDef": "<b>apple means</b> Round fruit with red or green skin.",
      },
      "answer": [
        "apple",
        "appel",
        "aple",
      ],
      "question": [
        "https://example.com/audio/apple.mp3",
      ],
      "wordId": "1",
    },
    {
      "additionalInfo": {
        "examples": [],
        "imgUrl": "https://example.com/images/orange.jpg",
        "shortDef": "<b>orange means</b> Citrus fruit with orange rind.",
      },
      "answer": [
        "orange",
        "oranj",
        "orenge",
      ],
      "question": [
        "https://example.com/audio/orange.mp3",
      ],
      "wordId": "2",
    },
    {
      "additionalInfo": {
        "examples": [],
        "imgUrl": "https://example.com/images/banana.jpg",
        "shortDef": "<b>banana means</b> Curved fruit with yellow skin.",
      },
      "answer": [
        "banana",
      ],
      "question": [
        "https://example.com/audio/banana.mp3",
      ],
      "wordId": "3",
    },
    {
      "additionalInfo": {
        "examples": [],
        "imgUrl": "https://example.com/images/strawberry.jpg",
        "shortDef": "<b>strawberry means</b> Soft red fruit with seeds.",
      },
      "answer": [
        "strawberry",
        "strawbery",
        "straw-berry",
      ],
      "question": [
        "https://example.com/audio/strawberry.mp3",
      ],
      "wordId": "4",
    },
    {
      "additionalInfo": {
        "examples": [],
        "imgUrl": "https://example.com/images/pineapple.jpg",
        "shortDef": "<b>pineapple means</b> Tropical fruit with spiky rind.",
      },
      "answer": [
        "pineapple",
      ],
      "question": [
        "https://example.com/audio/pineapple.mp3",
      ],
      "wordId": "5",
    },
    {
      "additionalInfo": {
        "examples": [],
        "imgUrl": "https://example.com/images/grape.jpg",
        "shortDef": "<b>grape means</b> Small juicy fruit with smooth skin.",
      },
      "answer": [
        "grape",
        "grap",
        "greyp",
      ],
      "question": [
        "https://example.com/audio/grape.mp3",
      ],
      "wordId": "6",
    },
  ],
  "task": "Type the word that you've heard",
  "type": "AUDIO",
}
`;

exports[`generateAudioGame should return the correct number of questions when data has enough SPANISH words 1`] = `
{
  "questions": [
    {
      "additionalInfo": {
        "examples": [
          {
            "text": "Ella comió una manzana roja para el desayuno.",
            "translation": "She ate a red apple for breakfast.",
          },
        ],
        "imgUrl": "https://example.com/images/manzana.jpg",
        "shortDef": "<b>manzana means</b> Round fruit with red or green skin.",
      },
      "answer": [
        "manzana",
        "manzana",
        "manzána",
      ],
      "question": [
        "https://example.com/audio/manzana.mp3",
      ],
      "wordId": "1",
    },
    {
      "additionalInfo": {
        "examples": [
          {
            "text": "Ella bebió jugo de naranja fresco esta mañana.",
            "translation": "She drank fresh orange juice this morning.",
          },
          {
            "text": "Llamaban la naranja “fruta del sol” por su color.",
            "translation": "They called the orange "sun fruit" because of its color.",
          },
        ],
        "imgUrl": "https://example.com/images/naranja.jpg",
        "shortDef": "<b>naranja means</b> Citrus fruit with orange rind.",
      },
      "answer": [
        "naranja",
        "naranga",
        "naranj",
      ],
      "question": [
        "https://example.com/audio/naranja.mp3",
      ],
      "wordId": "2",
    },
    {
      "additionalInfo": {
        "examples": [
          {
            "text": "Para evitar que se pegue, es mejor engrasar la sartén antes de freír.",
            "translation": "To prevent sticking, it's best to powder the pan before frying.",
          },
        ],
        "imgUrl": "https://example.com/images/skillet.jpg",
        "shortDef": "<b>sartén means</b> skillet",
      },
      "answer": [
        "sartén",
      ],
      "question": [
        "https://example.com/audio/platano.mp3",
      ],
      "wordId": "3",
    },
    {
      "additionalInfo": {
        "examples": [
          {
            "text": "Ella recogió una fresa madura del jardín.",
            "translation": "She picked a ripe strawberry from the garden.",
          },
        ],
        "imgUrl": "https://example.com/images/fresa.jpg",
        "shortDef": "<b>fresa means</b> Soft red fruit with seeds.",
      },
      "answer": [
        "fresa",
        "freza",
        "fre-sá",
      ],
      "question": [
        "https://example.com/audio/fresa.mp3",
      ],
      "wordId": "4",
    },
    {
      "additionalInfo": {
        "examples": [
          {
            "text": "Hicieron un batido de piña.",
            "translation": "They made a pineapple smoothie.",
          },
        ],
        "imgUrl": "https://example.com/images/pina.jpg",
        "shortDef": "<b>piña means</b> Tropical fruit with spiky rind.",
      },
      "answer": [
        "piña",
        "pina",
        "pi-ñá",
      ],
      "question": [
        "https://example.com/audio/pina.mp3",
      ],
      "wordId": "5",
    },
    {
      "additionalInfo": {
        "examples": [
          {
            "text": "Aproveche el servicio de translado de la más alta calidad.",
            "translation": "ake advantage of the highest quality transfer service.",
          },
        ],
        "imgUrl": "https://example.com/images/translado.jpg",
        "shortDef": "<b>translado means</b> transfer.",
      },
      "answer": [
        "translado",
      ],
      "question": [
        "https://example.com/audio/translado.mp3",
      ],
      "wordId": "6",
    },
  ],
  "task": "Type the word that you've heard",
  "type": "AUDIO",
}
`;

exports[`generateAudioGame should return the correct number of questions when with minimal ENGLISH words 1`] = `
{
  "questions": [
    {
      "additionalInfo": {
        "examples": [],
        "imgUrl": "https://example.com/images/apple.jpg",
        "shortDef": "<b>apple means</b> Round fruit with red or green skin.",
      },
      "answer": [
        "apple",
        "appel",
        "aple",
      ],
      "question": [
        "https://example.com/audio/apple.mp3",
      ],
      "wordId": "1",
    },
  ],
  "task": "Type the word that you've heard",
  "type": "AUDIO",
}
`;

exports[`generateAudioGame should return the correct number of questions when with minimal SPANISH words 1`] = `
{
  "questions": [
    {
      "additionalInfo": {
        "examples": [
          {
            "text": "Ella comió una manzana roja para el desayuno.",
            "translation": "She ate a red apple for breakfast.",
          },
        ],
        "imgUrl": "https://example.com/images/manzana.jpg",
        "shortDef": "<b>manzana means</b> Round fruit with red or green skin.",
      },
      "answer": [
        "manzana",
        "manzana",
        "manzána",
      ],
      "question": [
        "https://example.com/audio/manzana.mp3",
      ],
      "wordId": "1",
    },
  ],
  "task": "Type the word that you've heard",
  "type": "AUDIO",
}
`;
